### DISCUSSION TITLE
tar_map over tibble containing quosures

### DISCUSSION THREAD

**Discussion by djbirke** [ropensci/targets#435] [Help] [answered]
I am encountering an issue when applying `tar_map` to a tibble that contains quosures. I am unsure what do about it. Would someone have any ideas what is causing the issue and how I can make the following work?

``` r
library(targets)
library(tidyverse)
library(tarchetypes)

mean_plus_x <- function(group_vars, x) {
  mtcars %>%
    group_by(!!!group_vars) %>%
    summarize(mean(mpg) + 1, .groups = "drop")
}

list(
  tar_map(
    values = tidyr::expand_grid(
      x = c(0, 1, 2),
      group_vars = list(
        dplyr::vars(cyl),
        dplyr::vars(cyl, gear)
      )
    ),
    tar_target(summary, mean_plus_x(group_vars, x)),
    names = c(group_vars, x)
  )
)
#> Error in (function (..., row.names = NULL, check.rows = FALSE, check.names = TRUE, : arguments imply differing number of rows: 9, 6
```

<sup>Created on 2021-04-26 by the [reprex package](https://reprex.tidyverse.org) (v1.0.0)</sup>

**Comment by ginolhac** [comment]
Hello,

maybe "hello" and "please" would be appreciated here don't you think?

**Reply by djbirke** [reply]
Hi @ginolhac, thanks for the feedback. I don't think those exact words matter much—they are often absent in online Q&As, even in this repo—but I have extended my original post to avoid it being perceived as impolite or entitled to an answer. I am grateful to @wlandau for his answer here and his work on drake and target more broadly.

**Reply by ginolhac** [reply]
Thanks for your message. Maybe @wlandau does not mind and your description of lack of politeness online is indeed the norm. However I think that's a mistake for human relationship and for developers who work tremendously for free. William's work is amazing and I won't stop repeating and acknowledging it

**Reply by wlandau** [reply]
Thanks @ginolhac, I really appreciate the sentiment, especially given the rough weekend (#434).

@djbirke, no worries about any accidental abruptness. Your code was readable, concise, and complete, so it did not take me much time or effort to diagnose.

**Comment by wlandau** [comment, marked as answer]
Hi @djbirke,

In `tar_map()`, the `names` argument refers to columns of `values` that can be deparsed to suffixes of target names. `group_vars` is difficult to deparse, and the result is not usable for target names.

```r
deparse(list(vars(cyl), vars(cyl, gear)))
#> [1] "list(structure(list(~cyl), .Names = \"\", class = c(\"quosures\", "                       
#> [2] "\"list\")), structure(list(~cyl, ~gear), .Names = c(\"\", \"\"), class = c(\"quosures\", "
#> [3] "\"list\")))"      
```

So I recommend you create another column for the names.

``` r
library(targets)
tar_script({
  options(tidyverse.quiet = TRUE)
  library(tarchetypes)
  library(tidyverse)
  values <- expand_grid(
    x = c(0, 1, 2),
    tibble(
      group_vars = list(vars(cyl), vars(cyl, gear)),
      group_names = c("cyl", "cyl_gear")
    )
  )
  list(
    tar_map(
      values = values,
      names = all_of(c("x", "group_names")), # Omits group_vars.
      tar_target(summary, mean_plus_x(group_vars, x))
    )
  )
})

tar_manifest()
#> # A tibble: 6 x 3
#>   name               command                           pattern
#>   <chr>              <chr>                             <chr>  
#> 1 summary_2_cyl      mean_plus_x(list(~cyl), 2)        <NA>   
#> 2 summary_1_cyl      mean_plus_x(list(~cyl), 1)        <NA>   
#> 3 summary_2_cyl_gear mean_plus_x(list(~cyl, ~gear), 2) <NA>   
#> 4 summary_0_cyl      mean_plus_x(list(~cyl), 0)        <NA>   
#> 5 summary_1_cyl_gear mean_plus_x(list(~cyl, ~gear), 1) <NA>   
#> 6 summary_0_cyl_gear mean_plus_x(list(~cyl, ~gear), 0) <NA>
```

<sup>Created on 2021-04-27 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Reply by djbirke** [reply]
Great, thank you for the explanation and recommendation!
